File: src/main/kotlin/org/skialin/RenderNodeContext.kt
package org.skialin

import org.skialin.impl.Managed
import org.skialin.impl.NativeLoader

class RenderNodeContext(
    measureDrawBounds: Boolean = false,
    snapshotCache: Boolean = true,
) : Managed(RenderNodeContextNative.nMake(measureDrawBounds, snapshotCache), RenderNodeContextNative::nRelease) {
    fun setLightingInfo(
        centerX: Float = 0f,
        centerY: Float = 0f,
        centerZ: Float = 0f,
        radius: Float = 0f,
        ambientShadowAlpha: Float = 0f,
        spotShadowAlpha: Float = 0f,
    ) {
        RenderNodeContextNative.nSetLightingInfo(nativePtr, centerX, centerY, centerZ, radius, ambientShadowAlpha, spotShadowAlpha)
    }
}

private object RenderNodeContextNative {
    init {
        NativeLoader.ensureLoaded()
    }

    external fun nMake(
        measureDrawBounds: Boolean,
        snapshotCache: Boolean,
    ): Long

    external fun nRelease(ptr: Long)

    external fun nSetLightingInfo(
        ptr: Long,
        centerX: Float,
        centerY: Float,
        centerZ: Float,
        radius: Float,
        ambientShadowAlpha: Float,
        spotShadowAlpha: Float,
    )
}

File: native-shim/src/node/RenderNodeContext.cpp
#include "skialin/bridge.h"
#include "skialin/node/RenderNodeContext.h"

namespace skialin {
namespace node {

RenderNodeContext::RenderNodeContext(bool measureDrawBounds, bool snapshotCache)
    : lightGeometry{}, lightInfo{}, measureDrawBounds(measureDrawBounds), snapshotCache(snapshotCache) {}

void RenderNodeContext::setLightingInfo(const LightGeometry& lightGeometry, const LightInfo& lightInfo) {
    this->lightGeometry = lightGeometry;
    this->lightInfo = lightInfo;
}

} // namespace node
} // namespace skialin

skialin::node::RenderNodeContext* skialin_bridge_RenderNodeContext_Make(bool measureDrawBounds, bool snapshotCache) {
    return new skialin::node::RenderNodeContext(measureDrawBounds, snapshotCache);
}

void skialin_bridge_RenderNodeContext_unref(skialin::node::RenderNodeContext* context) {
    SkSafeUnref(context);
}

void skialin_bridge_RenderNodeContext_setLightingInfo(
    skialin::node::RenderNodeContext* context,
    float centerX, float centerY, float centerZ, float radius,
    float ambientShadowAlpha, float spotShadowAlpha) {
    skialin::node::LightGeometry lightGeometry{SkPoint3::Make(centerX, centerY, centerZ), radius};
    skialin::node::LightInfo lightInfo{ambientShadowAlpha, spotShadowAlpha};
    context->setLightingInfo(lightGeometry, lightInfo);
}
